src/ext.rs

```diff
@@ -0,0 +1,64 @@
+//! HTTP extensions
+
+use bytes::Bytes;
+#[cfg(feature = "http1")]
+use http::header::{GetAll, HeaderName, IntoHeaderName};
+use http::HeaderMap;
+
+/// A map from header names to their original casing as received in an HTTP message.
+///
+/// If an HTTP/1 response `res` is parsed on a connection whose option
+/// [`http1_preserve_header_case`] was set to true and the response included
+/// the following headers:
+///
+/// ```ignore
+/// x-Bread: Baguette
+/// X-BREAD: Pain
+/// x-bread: Ficelle
+/// ```
+///
+/// Then `res.extensions().get::<HeaderCaseMap>()` will return a map with:
+///
+/// ```ignore
+/// HeaderCaseMap({
+///     "x-bread": ["x-Bread", "X-BREAD", "x-bread"],
+/// })
+/// ```
+///
+/// [`http1_preserve_header_case`]: /client/struct.Client.html#method.http1_preserve_header_case
+#[derive(Clone, Debug, Default)]
+pub struct HeaderCaseMap(HeaderMap<RawHeaderName>);
+
+#[cfg(feature = "http1")]
+impl HeaderCaseMap {
+    /// Returns a view of all spellings associated with that header name,
+    /// in the order they were found.
+    pub fn get_all(&self, name: &HeaderName) -> GetAll<'_, RawHeaderName> {
```

Comment:
Just realized that in a way, this reveals that the internal type is a `HeaderMap`. Maybe that's fine eventually, but to not leak that, we could change the return type to `-> impl Iterator<Item = Blah>`. Any concern with that?

Reply:
Nah that's a good idea.